tracer/src/Datadog.Trace/Ci/TestModule.cs

```diff
@@ -0,0 +1,315 @@
+            LibraryVersion = TracerConstants.AssemblyVersion,
+            RuntimeName = frameworkDescription.Name,
+            RuntimeVersion = frameworkDescription.ProductVersion,
+            RuntimeArchitecture = frameworkDescription.ProcessArchitecture,
+            OSArchitecture = frameworkDescription.OSArchitecture,
+            OSPlatform = frameworkDescription.OSPlatform,
+            OSVersion = Environment.OSVersion.VersionString,
+        };
+
+        if (environment.VariablesToBypass is { } variablesToBypass)
+        {
+            tags.CiEnvVars = Vendors.Newtonsoft.Json.JsonConvert.SerializeObject(variablesToBypass);
+        }
+
+        // Check if Intelligent Test Runner has skippable tests
+        if (CIVisibility.HasSkippableTests())
+        {
+            tags.TestsSkipped = "true";
+        }
+
+        var span = Tracer.Instance.StartSpan(
+            string.IsNullOrEmpty(framework) ? "test_module" : $"{framework!.ToLowerInvariant()}.test_module",
+            tags: tags,
+            startTime: startDate);
+
+        span.Type = SpanTypes.TestModule;
+        span.ResourceName = name;
+        span.SetTag(Trace.Tags.Origin, TestTags.CIAppTestOriginName);
+        span.SetTraceSamplingPriority(SamplingPriority.AutoKeep);
+
+        tags.ModuleId = span.SpanId;
+
+        _span = span;
+        Current = this;
+        CIVisibility.Log.Debug("### Test Module Created: {name}", name);
+
+        if (startDate is null)
+        {
+            // If a module doesn't have a fixed start time we reset it before running code
+            span.ResetStartTime();
+        }
+    }
+
+    /// <summary>
+    /// Gets the current TestModule
+    /// </summary>
+    public static TestModule? Current
+    {
+        get => CurrentModule.Value;
+        internal set => CurrentModule.Value = value;
+    }
+
+    /// <summary>
+    /// Gets the module name
+    /// </summary>
+    public string Name { get; }
+
+    /// <summary>
+    /// Gets the test module start date
+    /// </summary>
+    public DateTimeOffset StartTime => _span.StartTime;
+
+    /// <summary>
+    /// Gets the test framework
+    /// </summary>
+    public string? Framework { get; }
+
+    internal TestModuleSpanTags Tags => (TestModuleSpanTags)_span.Tags;
+
+    /// <summary>
+    /// Create a new Test Module
+    /// </summary>
+    /// <param name="name">Test module name</param>
+    /// <returns>New test session instance</returns>
+    public static TestModule Create(string name)
+    {
+        return new TestModule(name, null, null, null);
+    }
+
+    /// <summary>
+    /// Create a new Test Module
+    /// </summary>
+    /// <param name="name">Test module name</param>
+    /// <param name="framework">Testing framework name</param>
+    /// <param name="frameworkVersion">Testing framework version</param>
+    /// <returns>New test session instance</returns>
+    public static TestModule Create(string name, string? framework, string? frameworkVersion)
```

Comment:
Can we make the `framework` parameter and the `Framework` property non-nullable?

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/commit/1804c381ada9ab0c3c2fcbab2e47173544f57df9